---
feature_text: |
  ## Genomic Visualization and Interpretations
title: Introduction to Genomic Data Visualization and Interpretation
categories:
    - Module 1
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0001-01-01
---

Welcome to the Genomic Visualization and Interpretations tutorial series. The goal of this tutorial is to provide students with the skills necessary to visualize and interpret 'omic' data in various forms. To accomplish this task we will be exploring a number of tools and resources available through the R programming language and via the web, providing background where appropriate. We recommend using the navigation links found at the bottom of all course topic pages however students can also search for individual topics via the search tab. Exercises to reinforce concepts are interspersed throughout each course topic. By the end of this course students should:

* Have a solid understanding of basic computational biology techniques for interpreting data
* Be familiar with a variety of web resources (e.g., genome browsers and portals) to aid in the interpretation of omic data
* Be able to use R graphical programs for data visualization (e.g., ggplot2, GenVisR) and interactive visualization (e.g., shiny)
* Be able to perform, visualize, and interpret expression, differential expression and pathway analysis
* Be able to assess a variant's biological and clinical significance (e.g., VEP, ClinVar, and CIViC)
