---
feature_text: |
  ## Genomic Visualization and Interpretations
title: ClinVar
categories:
    - Module 5
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0005-03-01
---

[ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/) aggregates information about genomic variation and its relationship to human health. To date, thousands of variants and associated phenotypes have been deposited in ClinVar. The main use case for ClinVar is when you have one or more variants observed in a human sample and you would like to know if these variants have established clinical relevance. The resource is fairly heavily biased towards germline inherited variants that predispose an individual for disease. Many diseases and phenotypes have been reported. Some of the variant observations in ClinVar come from sources that mine published literature (e.g. [DoCM](http://docm.info), [OMiM](https://www.omim.org/), etc.). However, the majority of submissions to ClinVar come from clinical sequencing labs that identify variants in patients and report that variant along with the phenotype of the patient that instigated testing (see [Data Sources](https://www.ncbi.nlm.nih.gov/clinvar/docs/datasources/) for more details). Several types of variants are supported in ClinVar. Many are single nucleotide variants (SNVs) or small insertions and deletions, but large structural variants are also supported. In general, all variants in ClinVar are designated using the HGVS nomenclature. 

### Basic intro to the ClinVar interface
ClinVar offers a few video tutorials to get you started: [Intro to ClinVar](https://youtu.be/A8G3ej83ZgU) and [Find All Variants with ClinVar](https://youtu.be/H09-0pP48Us). The ClinVar website is quite simple. The [home page](https://www.ncbi.nlm.nih.gov/clinvar/) offers various links to frequently used sections. The tool bar at the top of each page includes the simple search bar and various menus to navigate to additional information.

* To get a sense of where the data in ClinVar is coming from, visit the complete [submitter list](https://www.ncbi.nlm.nih.gov/clinvar/docs/submitter_list/) (Statistics -> List of Submitters). This list gives you an idea of some of the major players in the genetic diagnostics and variant interpretation.
* To get a sense of the current content of ClinVar, visit the [Statistics page](https://www.ncbi.nlm.nih.gov/clinvar/submitters/) (Statistics -> Statistics).
* If you would like to download raw ClinVar data, you can do so from their [FTP site](ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/). For example, you can download the complete set of ClinVar variants in VCF format ([human genome build 37 VCF](ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/) or [human genome build 38 VCF](ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/)).

ClinVar has several key entities used to organize variant data. (1) Each submission of an interpretation for a variant is assigned a **submission accession** of the format *SCV000000000.0*. Multiple labs could report independent interpretations for the same variant. Each would be entereed as a distinct record. If a lab wishes to update that submission, they can do so, and the version number will be updated to reflect this change. (2) When there are multiple submissions about the same variation/condition relationship, they are aggregated under a **reference accession** of the format *RCV000000000.0*. If the same variant is associated with multiple conditions, there will be multiple reference accessions. (3) All information about a single distinct variant are organized under one Variation ID and a corresponding Allele ID. Both are simple integers. Each allele corresponds to a distinct genomic variation that must be described using one or more HGVS expressions.

Consider the following example: [NM_007294.3(BRCA1):c.5558dupA (p.Tyr1853Terfs)](https://www.ncbi.nlm.nih.gov/clinvar/variation/55628/)

Variation ID: 55628
Allele ID: 70295
Submission accessions: SCV000300281.2, SCV000329130.4, SCV000488476.1, ...
Reference accessions: RCV000074357 (Breast-ovarian cancer, familial 1), ..., RCV000163976 (Hereditary cancer-predisposing syndrome)

### Searching ClinVar


### A brief primer on HGVS


### A brief primer on the ACMG guidelines for germline variant interpretation


### ClinVar practice exercises

How many ClinVar variants are there for BRCA2 that are Germline, Pathogenic and have Expert Panel review status?
{% include question.html question="Get a hint!" answer='Simply enter BRCA2 in the ClinVar search box, then apply the three filters requested.'%}
{% include question.html question="Answer" answer='At the time of writing this post, there were 2069 germline, pathogenic, expert panel reviewed BRCA2 variants.'%}

Use ClinVar to find valid HGVS expressions at the genome, cDNA, and protein level for the following variant: *PIK3CA* H1047R.
{% include question.html question="Get a hint!" answer='Simply enter PIK3CA H1047R in the ClinVar search box, then review the HGVS section.'%}
{% include question.html question="Answer" answer='NG_012113.2:g.90775A>G (genome), NM_006218.3:c.3140A>G (cDNA), NP_006209.2:p.His1047Arg (protein).'%}

Find all of the variants in ClinVar that correspond to exon 2 of VHL. How many are there? Export these to a spreadsheet.
{% include question.html question="Get a hint!" answer='Use a genome browser such as IGV to obtain the build38 coordinates for exon 2 of VHL. Then use the advanced ClinVar search box to search for variants in VHL and within the exon 2 coordinates.'%}
{% include question.html question="Answer" answer='To get VHL exon 2 variants use this query. (VHL[Gene Name]) AND 10146514:10146636[Base Position]. At the time this question was created there were 517 variants in VHL and 63 of these were within exon 2.'%}

