---
feature_text: |
  ## Genomic Visualization and Interpretations
title: ClinVar
categories:
    - Module 5
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0005-03-01
---

Brief introduction to ClinVar

[ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/) aggregates information about genomic variation and its relationship to human health. To date, thousands of variants and associated phenotypes have been deposited in ClinVar. The main use case for ClinVar is when you have one or more variants observed in a human sample and you would like to know if these variants have established clinical relevance. The resource is fairly heavily biased towards germline inherited variants that predispose an individual for disease. Many diseases and phenotypes have been reported. Some of the variant observations in ClinVar come from sources that mine published literature (e.g. [DoCM](http://docm.info), [OMiM](https://www.omim.org/), etc.). However, the majority of submissions to ClinVar come from clinical sequencing labs that identify variants in patients and report that variant along with the phenotype of the patient that instigated testing. Several types of variants are supported in ClinVar. Many are single nucleotide variants (SNVs) or small insertions and deletions, but large structural variants are also supported. In general, all variants in ClinVar are designated using the HGVS nomenclature. 

### Basic intro to the ClinVar interface


### Searching ClinVar


### An aside on HGVS


### ClinVar practice exercises

Question text?
{% include question.html question="Get a hint!" answer='hint text'%}
{% include question.html question="Answer" answer='answer text'%}
